### PR TITLE
Mutation types

### PR DESCRIPTION
**Pull request by Kelenius**
#### Summary
```SUMMARY: Infrastructure "Mutation types added; player can't have two mutations of the same type."```

#### Purpose of change
Mutations are generally fairly messy, and how they cancel each other is one of the messy parts. Errors like #24156 and #24124 are easy to make, and mutation mods are hard to make compatible with each other.

#### Describe the solution
Mutation types are a simple tool - each mutation can belong to one or several types, and the character can't have two mutations of the same type. This means that, if you have a large amount of mutations cancelling each other, you don't need to list them individually, you just need to give them the same type.

#### Additional context
It's WIP because I'm still scratching my head at the mutations that affect your face, and I want to test it some more.

I'm also seriously considering burning the mutation code to the ground and rewriting it from the scratch. It's a pain to work with.

### CODE REVIEW

File: src/mutation_type.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+std::vector<trait_id> get_mutations_in_types( std::set<std::string> ids )
+{
+    std::vector<trait_id> ret;
+    for( auto it : ids ) {
+        std::vector<trait_id> this_id = get_mutations_in_type( it );
+        ret.insert( ret.end(), this_id.begin(), this_id.end() );
+    }
+    return ret;
+}
```
Comment (ZhilkinSerg) [inline src/mutation_type.cpp]: Missing linebreak

**Comment by Night-Pryanik** [comment]
> It's WIP because I'm still scratching my head at the mutations that affect your face, and I want to test it some more.

I'd add `VISION` or `EYES` type for all types of eyes - feline, fey, night vision, IR-vision and the like. Though this will prevent us having both far- and nearsightedness at the same time.

Also something like `FOOD_PREFERENCE` for vegetarian, meatarian and the like. Also maybe carnivore vs herbivore and all their descendants.

**Comment by kevingranade** [comment]
Please only add a few types along with the feature.  Replacing the existing rules with this sounds like a great cleanup, but I dont want to review dozens of different mutation interactions in one PR.

**Comment by Xhuis** [comment]
You will likely need to update this when #24672 is merged (or I will need to update it when this is merged.) As far as I know, Mouse is ready.

**Comment by cainiaowu** [comment]
Well, all those documents need some update too.

**Comment by Pixelmage** [comment]
> I'd add VISION or EYES type for all types of eyes - feline, fey, night vision, IR-vision and the like. Though this will prevent us having both far- and nearsightedness at the same time.

The simple solution would be to add one new negative trait; Poor Vision, -4 points, counter with Bifocals, has the combined effects of Far and Near sighted. That way we preserve the way things work now, while also allowing for the under-the-hood advantage of unique exclusive categories.

**Comment by Kelenius** [comment]
Alternating between slime and chimera mutagen correctly moves you between regeneration and deterioration, and doing it with fish and plant moves between bark and sleek scales. Everything should be in order, more types will be added in a different PR.

**Comment by narc0tiq** [comment]
Sorry, only commenting for the semi-obscure joke, but it sounds like these are mutation exclusion groups, aka _mutex groups_? :)

Also, yes, good thought here.

**Comment by narc0tiq** [comment]
Jenkins, rebuild please.

This comment will self-destruct; testing 9ab1f210c6618b12fde9dbe105f179f508420a70.

File: src/mutation.h [hunk 2]
```diff
@@ -319,9 +320,12 @@ struct mutation_category_trait {
     static void check_consistency();
 };
 
+void load_mutation_type( JsonObject &jsobj );
 void load_mutation_category( JsonObject &jsobj );
 void load_dream( JsonObject &jsobj );
 bool mutation_category_is_valid( const std::string &cat );
+bool mutation_type_exists( std::string id );
```
Comment (ZhilkinSerg) [inline src/mutation.h]: Why not const reference?

File: src/mutation_data.cpp [hunk 3]
```diff
@@ -378,6 +379,12 @@ void mutation_branch::check_consistency()
                 debugmsg( "mutation %s refers to undefined martial art style %s", mid.c_str(), style.c_str() );
             }
         }
+        for( std::string type : mdata.types )
```
Comment (ZhilkinSerg) [inline src/mutation_data.cpp]: Why not `const std::string` or `const auto`?

File: src/mutation_type.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+void load_mutation_type( JsonObject &jsobj )
+{
+    mutation_type new_type;
+    new_type.id = jsobj.get_string( "id" );
+
+    mutation_types[new_type.id] = new_type;
+}
+
+bool mutation_type_exists( std::string id )
```
Comment (ZhilkinSerg) [inline src/mutation_type.cpp]: const reference

File: src/mutation_type.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+{
+    return mutation_types.find( id ) != mutation_types.end();
+}
+
+std::vector<trait_id> get_mutations_in_type( std::string id )
```
Comment (ZhilkinSerg) [inline src/mutation_type.cpp]: const reference

File: src/mutation_type.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    std::vector<trait_id> ret;
+    for( auto it : mutation_branch::get_all() ) {
+        if( it.second.types.find( id ) != it.second.types.end() ) {
+            ret.push_back( it.first );
+        }
+    }
+    return ret;
+}
+
+std::vector<trait_id> get_mutations_in_types( std::set<std::string> ids )
```
Comment (ZhilkinSerg) [inline src/mutation_type.cpp]: const reference

Comment (Kelenius) [reply src/mutation_type.cpp]: I don't think this one can be done that way?..

**Pull request merged**